### PR TITLE
RIO - pyocf IO excersizer

### PR DESCRIPTION
**Pull request by jfckm**
Introducing rio(Rigid IO tester) - simple tool for quickly testing OCF cache IO in pyocf.

### CODE REVIEW

**Comment by jfckm** [comment]
retest this please

**Comment by KamilLepek** [comment]
retest this please

**Comment by KamilLepek** [comment]
retest this please

File: tests/functional/tests/engine/test_pp.py [hunk 1]
```diff
@@ -62,49 +64,35 @@ def test_change_to_nhit_and_back_io_in_flight(pyocf_ctx):
...
+    )
 
     # Step 3
     cache.set_promotion_policy(PromotionPolicy.NHIT)
 
     # Step 4
-    for c in completions:
-        c.wait()
-        assert not c.results["error"], "No IO's should fail when turning NHIT policy on"
+    r.abort()
```
Comment (arutk) [inline tests/functional/tests/engine/test_pp.py:84]: I think you mean "wait()" rather than "abort()" here - maybe add wait() to the Rio class?

**Review by arutk — APPROVED** [review]

File: tests/functional/pyocf/rio.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        return other
+
+
+class Rio:
+    class RioThread:
+        pass
+
+
+class Rio:
+    class RioThread(Thread):
```
Comment (rafalste) [inline tests/functional/pyocf/rio.py:86]: Does it have any purpose to define this class twice or is it just an oversight?

File: tests/functional/pyocf/rio.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    def run(self):
+        self.run_async()
+        self.wait_for_completion()
+        return self
+
+    def run_async(self):
+        self.clear()
+
+        jobs = self.jobs
```
Comment (rafalste) [inline tests/functional/pyocf/rio.py:277]: I don't see any reason for this assignment. Did you want to copy self.jobs to not modify the original?

File: tests/functional/pyocf/rio.py (hunk 2 above)
Comment (jfckm) [reply tests/functional/pyocf/rio.py]: Whoops, that wasn't supposed to be here

**Review by arutk — DISMISSED** [review]

File: tests/functional/pyocf/rio.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        return self
+
+    def run_async(self):
+        self.clear()
+
+        jobs = deepcopy(self.jobs)
+
+        if not jobs:
+            for _ in range(self.global_jobspec.njobs):
+                jobs.append(self.global_jobspec)
```
Comment (mmichal10) [inline tests/functional/pyocf/rio.py:277]: The loop seems unnecessary. How about `jobs = [self.global_jobspec] * self.global_jobspec.njobs`?

File: tests/functional/pyocf/rio.py (hunk 3 above)
Comment (jfckm) [reply tests/functional/pyocf/rio.py]: yup, I corrected it by using deepcopy to copy those jobspecs as well

**Review by arutk — APPROVED** [review]

File: tests/functional/pyocf/rio.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+            self.gen = cycle(gen)
+        else:
+            self.gen = gen
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.random and not self.randommap:
+            return choice(self.gen)
```
[the remark this answers is not in the document]
Comment (jfckm) [reply tests/functional/pyocf/rio.py]: this is to make norandommap behavior work - each time we want to pick from the same list of LBAs

**Comment by arutk** [comment]
retest this please

**Review by rafalste — APPROVED** [review]

**Comment by arutk** [comment]
this change introduces sporadic failures in functional tests and should not be integrated before this gets fixed

**Pull request closed without merging**